src/BenchmarkExamples/ForLoop/ForLoopBenchmark.cs

```diff
@@ -46,11 +46,42 @@ public void ForEachLoop()
         public void WithIndexLoop()
         {
             var sum = 0;
-            var intArray = new object[N];
             foreach (var (item, index) in intArray.WithIndex())
                 sum++;
         }
 
+        [Benchmark]
+        public void WithIndexCachedDelegateLoop()
+        {
+            var sum = 0;
+            foreach (var (item, index) in intArray.WithIndexCachedDelegate())
+                sum++;
+        }
+
+        [Benchmark]
+        public void WithIndexIteratorLoop()
+        {
+            var sum = 0;
+            foreach (var (item, index) in intArray.WithIndexIterator())
+                sum++;
+        }
+
+        [Benchmark]
+        public void WithIndexCustomEnumeratorLoop()
+        {
+            var sum = 0;
+            foreach (var (item, index) in intArray.WithIndexCustomEnumerator())
+                sum++;
+        }
+
+        [Benchmark]
+        public void WithIndexSpecificEnumeratorLoopArray()
+        {
+            var sum = 0;
+            foreach (var (item, index) in intArray.WithIndexArray())
+                sum++;
+        }
```

Comment:
This is the best I was able to achieve for an array. 13 to 18 times the baseline